File: MaiCheTong-V4.1.1/BuyCarDictionary/Index/ParameterCarTypeView.m
//
//  ParameterCarTypeView.m
//  BuyCarDictionary
//
//  Created by zgjun on 14-12-29.
//  Copyright (c) 2014年 chexun. All rights reserved.
//

#import "ParameterCarTypeView.h"
#import "NoHighLightBtn.h"
#import "UIImage+Extension.h"
#import "CompareButton.h"
#define GapWidth 15

@interface ParameterCarTypeView()
@property (nonatomic,weak) UILabel *carTypeName;
@property (nonatomic,weak) CompareButton *compareBtn;
@property (nonatomic,weak) UIImageView *imageView;
@property (nonatomic,weak) NoHighLightBtn *reduceBtn;

@property (nonatomic,strong) NSMutableArray *compareArrM;

@end

@implementation ParameterCarTypeView

- (NSMutableArray *)compareArrM {
    _compareArrM = [NSMutableArray array];
    
    //从沙盒里面读取历史数据
    NSString *docDir = [NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES) lastObject];
    //拼接文件名
    NSString *fileName = [docDir stringByAppendingPathComponent:@"compareCars.plist"];
    //先读取之前的历史记录
    _compareArrM = [[NSMutableArray alloc]initWithContentsOfFile:fileName];
    
    
    return _compareArrM;
}

- (instancetype)initWithFrame:(CGRect)frame {
    if (self = [super initWithFrame:frame]) {
        UIImageView *imageView = [[UIImageView alloc]init];
        self.imageView = imageView;
        imageView.userInteractionEnabled = YES;
        imageView.image = [UIImage resizableImageWithName:@"chexun_models_addbut_hengban"];
        [self addSubview:imageView];
        
        
        UILabel *carTypeName = [[UILabel alloc]init];
        carTypeName.font = [UIFont systemFontOfSize:12];
        carTypeName.numberOfLines = -1;
        carTypeName.textAlignment = NSTextAlignmentCenter;
        self.carTypeName = carTypeName;
        [self.imageView addSubview:carTypeName];
        
        CompareButton *compareBtn = [[CompareButton alloc]init];
        [compareBtn setImage:[UIImage imageNamed:@"chexun_pk_addicon"] forState:UIControlStateNormal];
        [compareBtn setImage:[UIImage imageNamed:@"chexun_models_cancelicon"] forState:UIControlStateSelected];
        //chexun_models_cancelicon
        [compareBtn setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
        [compareBtn setTitle:@"添加对比" forState:UIControlStateNormal];
        [compareBtn setTitle:@"取消对比" forState:UIControlStateSelected];
        compareBtn.imageView.contentMode = UIViewContentModeCenter;
        compareBtn.titleLabel.font = [UIFont systemFontOfSize:15];
        self.compareBtn = compareBtn;
        [compareBtn addTarget:self action:@selector(compareBtnClick:) forControlEvents:UIControlEventTouchUpInside];
        [self.imageView addSubview:compareBtn];
        
        NoHighLightBtn *reduceBtn = [[NoHighLightBtn alloc]init];
        [reduceBtn setBackgroundImage:[UIImage imageNamed:@"chexun_deletebut"] forState:UIControlStateNormal];
        self.reduceBtn = reduceBtn;
        [reduceBtn addTarget:self action:@selector(reduceBtnClick:) forControlEvents:UIControlEventTouchUpInside];
        
        [self addSubview:reduceBtn];
    }
    return self;
}

- (void)compareBtnClick:(NoHighLightBtn *)btn {
    
    NSMutableArray *compareTemp = [NSMutableArray arrayWithArray:self.compareArrM];
    
    if (compareTemp.count >= 10 && btn.selected == NO) {
        UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"亲，请注意" message:@"最多只能添加10款车进行对比选择" delegate:self cancelButtonTitle:@"确定" otherButtonTitles:nil];
        [alert show];
        return;
    }
    
    btn.selected = !btn.selected;
    
    if (btn.selected == YES) {
        //添加
        NSString *carTypeId = [NSString stringWithFormat:@"%@",self.dataDict[@"carTypeId"]];
        
        NSString *carTypeName = [NSString stringWithFormat:@"%@",self.dataDict[@"carTypeName"]];
        
        NSString *carTypeImage = [NSString stringWithFormat:@"%@",self.dataDict[@"carTypeImage"]];
        
        NSString *carTypePrice = [NSString stringWithFormat:@"%@",self.dataDict[@"carTypePrice"]];
        
        NSString *carTypeLowPrice = [NSString stringWithFormat:@"%@",self.dataDict[@"carTypeLowPrice"]];
        
        NSDictionary *carTypeInfo = @{@"carTypeId":carTypeId,@"carTypeName":carTypeName,@"carTypeImage":carTypeImage,@"carTypePrice":carTypePrice,@"carTypeLowPrice":carTypeLowPrice,@"compareState":@0,@"carSeriesId":self.dataDict[@"carSeriesId"],@"carSeriesName":self.dataDict[@"carSeriesName"]};
        
        [compareTemp addObject:carTypeInfo];
        
        if (compareTemp.count == 0) {
            //清除文件
            [self cleanCompareCars];
        } else {
            //写入新数据
            [self writeNewData:compareTemp];
        }
        
    } else {
        //减少
        for (int i = 0; i < compareTemp.count; i++) {
            
            NSDictionary *dict = compareTemp[i];
            
            if ([dict[@"carTypeId"] integerValue] == [self.dataDict[@"carTypeId"] integerValue]) {
                [compareTemp removeObject:dict];
                if (compareTemp.count == 0) {
                    //清除文件
                    [self cleanCompareCars];
                } else {
                    //写入新数据
                    [self writeNewData:compareTemp];
                }
                break;
            }
            
        }
    }
    
    //通知代理
    if ([self.delegate respondsToSelector:@selector(parameterAddBtnClick:)]) {
        [self.delegate parameterAddBtnClick:self];
    }
    
}

- (void)writeNewData:(NSMutableArray *)compareTemp {
    //从沙盒里面读取历史数据
    NSString *docDir = [NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES) lastObject];
    //拼接文件名
    NSString *fileName = [docDir stringByAppendingPathComponent:@"compareCars.plist"];
    
    [compareTemp writeToFile:fileName atomically:YES];
}

- (void)cleanCompareCars {
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
    NSString *documentsDirectory = [paths objectAtIndex:0];
    
    NSFileManager *fileMgr = [NSFileManager defaultManager];
    NSString *MapLayerDataPath = [documentsDirectory stringByAppendingPathComponent:@"compareCars.plist"];
    BOOL bRet = [fileMgr fileExistsAtPath:MapLayerDataPath];
    if (bRet) {
        NSError *err;
        [fileMgr removeItemAtPath:MapLayerDataPath error:&err];
    }
}
- (void)reduceBtnClick:(NoHighLightBtn *)btn {
    if ([self.delegate respondsToSelector:@selector(parameterReduceBtnClick:)]) {
        [self.delegate parameterReduceBtnClick:self];
    }
}

- (void)layoutSubviews {
    [super layoutSubviews];
    self.imageView.frame = CGRectMake(0, 10, self.width - GapWidth, self.height - 10);
    self.carTypeName.frame = CGRectMake(0, 0, self.imageView.width, self.imageView.height * 0.6);
    self.compareBtn.frame = CGRectMake(0, self.imageView.height * 0.6, self.imageView.width, self.imageView.height * 0.4);
    self.reduceBtn.frame = CGRectMake(self.width - 25, 0, 23, 23);
}

- (void)setDataDict:(NSDictionary *)dataDict {
    _dataDict = dataDict;
    
    
    self.carTypeName.text = dataDict[@"carTypeName"];
    
    CGFloat tipLabelX = 0;
    
    CGFloat tipLabelY = 0;
    
    CGSize maxSize = CGSizeMake(self.width, MAXFLOAT);
    
    CGSize tipLabelSize = [dataDict[@"value"] sizeWithFont:self.carTypeName.font constrainedToSize:maxSize];
    
    self.carTypeName.frame = (CGRect){{tipLabelX,tipLabelY},tipLabelSize};
    
    for (int i = 0; i < self.compareArrM.count; i++) {
        NSDictionary *dict = self.compareArrM[i];
        if ([dataDict[@"carTypeId"] integerValue] == [dict[@"carTypeId"] integerValue]) {
            self.compareBtn.selected = YES;
            break;
        }
    }
    
}

@end

File: GJScrollViewUsage_Swift/GJScrollViewUsage_Swift/View/ParameterCarTypeView.swift
//
//  ParameterCarTypeView.swift
//  GJScrollViewUsage_Swift
//
//  Created by zgjun on 15/9/22.
//  Copyright (c) 2015年 chexun. All rights reserved.
//

import UIKit

extension UIImage {
    //类方法
   class func resizableImageWithName(imageName:String) -> UIImage {
        var image : UIImage = UIImage(named: imageName)!
        var imageSize : CGSize = image.size
        image = image.resizableImageWithCapInsets(UIEdgeInsetsMake(imageSize.height * 0.4, imageSize.width * 0.5, imageSize.height * 0.6, imageSize.width * 0.5))
        
        return image
    }
}

//portocol
protocol ParameterCarTypeViewDelegate {
    func parameterReduceBtnClick (carTypeView : ParameterCarTypeView)
    func parameterAddBtnClick (carTypeView: ParameterCarTypeView)
}



class ParameterCarTypeView: UIView {
    
    var imageView :UIImageView!
    var carTypeName:UILabel!
    var compareBtn:CompareButton!
    var reduceBtn:NoHighLightBtn!
    //delegate
    var delegate : ParameterCarTypeViewDelegate!
    
    var compareArrM:NSMutableArray {
        get {
            
            if NSMutableArray(contentsOfFile: self.getCompareCarsPlistPath()) == nil {
                return NSMutableArray.new()
            } else {
                
                return NSMutableArray(contentsOfFile: self.getCompareCarsPlistPath())!
            }
            
        }
    }
    
    
    var dataDict :NSDictionary! {
        
        /**
        1.didSet wrong :fatal error: unexpectedly found nil while unwrapping an Optional value
        2.willSet was right
        */
        
        willSet(newDataDict) {
            let carType : AnyObject? = newDataDict["carTypeName"]
            carTypeName.text = "\(carType!)"
            let tipLabelX:CGFloat = 0
            let tipLabelY:CGFloat = 0
            let maxSize:CGSize = CGSizeMake(self.bounds.size.width, CGFloat(MAXFLOAT))
            
            var tipStrTemp : AnyObject!
            var tipStr :NSString!
            
            
            
            if newDataDict["value"] == nil {
                tipStr = ""
            } else {
                tipStrTemp = newDataDict["value"]!
                tipStr = "\(tipStrTemp)"
            }
            
            let tipLabelSize: CGSize = tipStr.sizeWithAttributes([NSFontAttributeName: UIFont.systemFontOfSize(14.0)])
            
            carTypeName.frame = CGRectMake(tipLabelX, tipLabelY, tipLabelSize.width, tipLabelSize.height)
            
            if compareArrM.count > 0 {
                
                for i in 0 ... compareArrM.count - 1 {
                    let dict : NSDictionary = compareArrM[i] as! NSDictionary
                    if newDataDict["carTypeId"]?.integerValue == dict["carTypeId"]?.integerValue {
                        compareBtn.selected = true
                        break
                    }
                }
            }
            
        }
    }
    /**
    get path of the compareCars.plist
    
    :returns: the  file's path
    */
    func getCompareCarsPlistPath () -> String {
        let docDir :NSArray = NSSearchPathForDirectoriesInDomains(.DocumentDirectory,.UserDomainMask, true) as NSArray
        var path :String = docDir.lastObject as! String
        path = path.stringByAppendingPathComponent("compareCars.plist")
        return path
    }
    
    

    override init(frame: CGRect) {
        super.init(frame: frame)
        
        imageView = UIImageView.new()
        imageView.userInteractionEnabled = true
        imageView.image = UIImage.resizableImageWithName("chexun_models_addbut_hengban")
        self.addSubview(imageView)
        
        carTypeName = UILabel.new()
        carTypeName.font = UIFont.systemFontOfSize(12)
        carTypeName.numberOfLines = -1
        carTypeName.textAlignment = NSTextAlignment.Center
        imageView.addSubview(carTypeName)
        
        compareBtn = CompareButton.new()
        compareBtn.setImage(UIImage(named: "chexun_pk_addicon"), forState: UIControlState.Normal)
        compareBtn.setImage(UIImage(named: "chexun_models_cancelicon"), forState: UIControlState.Selected)
        compareBtn.setTitleColor(UIColor.whiteColor(), forState: UIControlState.Normal)
        compareBtn.setTitle("添加对比", forState: UIControlState.Normal)
        compareBtn.setTitle("取消对比", forState: UIControlState.Selected)
        compareBtn.imageView!.contentMode = UIViewContentMode.Center
        compareBtn.contentVerticalAlignment = UIControlContentVerticalAlignment.Center
        compareBtn.titleLabel!.font = UIFont.systemFontOfSize(15)
        compareBtn.addTarget(self, action: "compareBtnClick:", forControlEvents: UIControlEvents.TouchUpInside)
        imageView.addSubview(compareBtn)
        
        reduceBtn = NoHighLightBtn.new()
        reduceBtn.setBackgroundImage(UIImage(named: "chexun_deletebut"), forState: UIControlState.Normal)
        reduceBtn.addTarget(self, action: "reduceBtnClick:", forControlEvents: UIControlEvents.TouchDown)
        self.addSubview(reduceBtn)
    }

    required init(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        let selfWidth : CGFloat = self.bounds.size.width
        let selfHeight : CGFloat = self.bounds.size.height
        let gapWidth : CGFloat = 15
        imageView.frame = CGRectMake(0, 10, selfWidth - gapWidth, selfHeight - 10)
        carTypeName.frame = CGRectMake(0, 0, selfWidth - gapWidth, selfHeight * 0.6)
        compareBtn.frame = CGRectMake(0, selfHeight * 0.55, selfWidth, selfHeight * 0.4)
        reduceBtn.frame = CGRectMake(selfWidth - 25, 0, 23, 23)
    }
    
    
    
    
    func compareBtnClick (btn : CompareButton) {
        
        btn.selected = !btn.selected
        
        var compareTemp :NSMutableArray = NSMutableArray(array: compareArrM)
        
        if btn.selected == true {
            let carTypeId : NSString = dataDict["carTypeId"] as! NSString
            let carTypeName : NSString = dataDict["carTypeName"] as! NSString
            let carTypeImage : NSString = dataDict["carTypeImage"] as! NSString
            let carTypePrice : NSString = dataDict["carTypePrice"] as! NSString
            let carTypeLowPrice :NSString = dataDict["carTypeLowPrice"] as! NSString
            let carSeriesId : NSString = dataDict["carSeriesId"] as! NSString
            let carSeriesName : NSString = dataDict["carSeriesName"] as! NSString
            
            var carTypeInfo : NSDictionary = ["carTypeId":carTypeId,"carTypeName":carTypeName,"carTypeImage":carTypeImage,"carTypePrice":carTypePrice,"carTypeLowPrice":carTypeLowPrice,"compareState":"0","carSeriesId":carSeriesId,"carSeriesName":carSeriesName]
            
            compareTemp.addObject(carTypeInfo)
            
            if compareTemp.count == 0 {
                self.cleanCompareCars()
            } else {
                self.writeNewData(compareTemp)
            }
            
        } else {
            for i in 0 ... compareTemp.count - 1 {
                var dict = compareTemp[i] as! NSDictionary
                if dict["carTypeId"]?.integerValue == dataDict["carTypeId"]?.integerValue {
                    compareTemp.removeObject(dict)
                    if compareTemp.count == 0 {
                        self.cleanCompareCars()
                    } else {
                        self.writeNewData(compareTemp)
                    }
                    break
                }
            }
        }
        
        //inform delegage do 
        delegate.parameterAddBtnClick(self)
        
    }
    
    func reduceBtnClick (btn :NoHighLightBtn) {
        delegate.parameterReduceBtnClick(self)
    }
    
    /**
    clear cars
    */
    func cleanCompareCars () {
        var path :String = getCompareCarsPlistPath()
        
        let fileMgr : NSFileManager = NSFileManager.defaultManager()
        var bRet = fileMgr.fileExistsAtPath(path)
        if bRet {
            var err:NSErrorPointer!
            fileMgr.removeItemAtPath(path, error: err)
        }
        
    }
    /**
    write to the file
    
    :param: arr cars
    */
    func writeNewData(arr: NSArray) {
        var path :String = getCompareCarsPlistPath()
        arr.writeToFile(path, atomically: true)
    }
}
